add consensus UI

dreacot opened the pull request:
Requires https://github.com/planetdecred/dcrlibwallet/pull/216

resolves #655
Fix #776

![Screenshot from 2022-01-21 14-59-46](https://user-images.githubusercontent.com/25265396/153193224-d8926d96-5f9c-4ada-a236-2146a755a365.png)
![Screenshot from 2022-01-21 15-00-01](https://user-images.githubusercontent.com/25265396/153193220-0d936c72-d7e7-4a34-a874-19b0a1f6ec3c.png)
![Screenshot from 2022-01-21 15-00-30](https://user-images.githubusercontent.com/25265396/153193210-8efbadfc-23f7-4a5f-ac78-20ac166aa1a2.png)

dreacot committed:
* add tab to governance page to switch between proposal and consensus pages
* implement consensus page

dreacot committed:
* add tab to governance page to switch between proposal and consensus pages
* implement consensus page

dreacot committed:
- add consensus listeners
- imlement vote for consensus
- cleanup code

dreacot committed:
- add consensus listeners
- imlement vote for consensus
- cleanup code

dreacot committed:
remove notifcation listemers implementations

dreacot committed:
remove notifcation listemers implementations

dreacot committed:
remove syncing section
* cleanup code

dreacot committed:
remove syncing section
* cleanup code

dreacot committed:
fix rebase error

dreacot committed:
fix rebase error

dreacot committed:
code cleanup

dreacot committed:
code cleanup

dreacot committed:
w.i.p reload page after voting

dreacot committed:
w.i.p reload page after voting

dreacot committed:
add dcrlibwallet PR requirement

dreacot committed:
add dcrlibwallet PR requirement

dreacot committed:
fix bug where agenda page wasn't reloaded after voting

dreacot committed:
fix bug where agenda page wasn't reloaded after voting

dreacot committed:
add ticket dropdown

dreacot committed:
add ticket dropdown

dreacot committed:
fix bug where ticket list wasn't gotten per wallet, but gotten for all wallets

dreacot committed:
fix bug where ticket list wasn't gotten per wallet, but gotten for all wallets

dreacot committed:
- fix error where pubkey was 0 length
- handle getting agenda status from dcrlibwallet

dreacot committed:
- fix error where pubkey was 0 length
- handle getting agenda status from dcrlibwallet

dreacot committed:
fix bug where the vote modal voting options wasn't updating based on the selected wallet

dreacot committed:
fix bug where the vote modal voting options wasn't updating based on the selected wallet

dreacot committed:
code cleanup

dreacot committed:
code cleanup

dreacot committed:
remove validation check for if a ticket is selected while trying to vote on agendas

dreacot committed:
use proper background color for agenda status

dreacot committed:
use proper background color for agenda status

dreacot committed:
update UI for voting modal

dreacot committed:
update UI for voting modal

dreacot committed:
- adjust governance page tab UI

dreacot committed:
- adjust governance page tab UI

dreacot committed:
code cleanup

dreacot committed:
code cleanup

dreacot committed:
update dcrlibwallet implementation

dreacot committed:
cleanup code

dreacot committed:
cleanup code

Sirmorrison commented on ui/load/vsp.go:
```diff
@@ -0,0 +1,193 @@
+package load
```
The file is no longer valid as all the details here have been migrated to dcrlibwallet. You will need to rebase your dcrlibwallet PR.

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,374 @@
+	"github.com/planetdecred/godcr/ui/decredmaterial"
+	"github.com/planetdecred/godcr/ui/load"
+	"github.com/planetdecred/godcr/ui/page/components"
+	"github.com/planetdecred/godcr/ui/page/staking"
```
This seems to be where the import cycle issue is coming from. To use the VSP selector, you could simply move it to the components.go file.

dreacot committed:
- move vsp selector to components file
- delete vsp.go file
- cleanup code

dreacot committed:
- move vsp selector to components file
- delete vsp.go file
- cleanup code

dreacot committed:
code cleanup

dreacot committed:
code cleanup

Sirmorrison commented on ui/load/load.go:
```diff
@@ -43,7 +43,7 @@ type Icons struct {
 	Restore, DocumentationIcon, DownloadIcon, TimerIcon, StakeIcon, StakeIconInactive, StakeyIcon,
 	List, ListGridIcon, DecredSymbolIcon, DecredSymbol2, GovernanceActiveIcon, GovernanceInactiveIcon,
-	LogoDarkMode, TimerDarkMode, Rebroadcast, SettingsActiveIcon, SettingsInactiveIcon *decredmaterial.Image
+	LogoDarkMode, TimerDarkMode, Rebroadcast, SettingsActiveIcon, SettingsInactiveIcon, ActivatedActiveIcon, ActivatedInactiveIcon, LockinActiveIcon, LockinInactiveIcon *decredmaterial.Image
```
```suggestion
	LogoDarkMode, TimerDarkMode, Rebroadcast, SettingsActiveIcon, SettingsInactiveIcon, ActivatedIcon, 
	DeactivatedIcon, LockinActiveIcon, LockinInactiveIcon *decredmaterial.Image
```

break it into a new line after `ActivatedActiveIcon`

Sirmorrison commented on ui/load/load.go:
```diff
@@ -184,6 +185,10 @@ func IconSet() Icons {
 		SettingsActiveIcon:       decredmaterial.NewImage(decredIcons["settings_active"]),
 		SettingsInactiveIcon:     decredmaterial.NewImage(decredIcons["settings_inactive"]),
+		ActivatedActiveIcon:      decredmaterial.NewImage(decredIcons["activated_active"]),
+		ActivatedInactiveIcon:    decredmaterial.NewImage(decredIcons["activated_inactive"]),
```
same as above this stutters.

dreacot replied:
![Screenshot from 2022-02-11 00-45-07](https://user-images.githubusercontent.com/25265396/153515946-23a30bf3-b9a3-4486-8f51-012489da0151.png)

The icon in itself is an acivated Icon(used same way on voting.decred.org), and the icon is in an activated state, the icon also has an inactive state

i'm open to suggestions, but as it is currently, the naming isn't from me

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,167 @@
+		canVote = true
+	case "Upcoming":
+		statusLabel = l.Theme.Label(values.MarginPadding14, agenda.Status)
+		statusLabelColor = l.Theme.Color.DeepBlue
```
```suggestion
		statusLabelColor = l.Theme.Color.Text
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,167 @@
+	switch agenda.Status {
+	case "Finished":
+		statusLabel = l.Theme.Label(values.MarginPadding14, agenda.Status)
+		statusLabelColor = l.Theme.Color.Success
```
```suggestion
		statusLabelColor = l.Theme.Color.GreenText
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,167 @@
+				}),
+				layout.Rigid(func(gtx C) D {
+					return statusLabel.Layout(gtx)
+				}))
```
```suggestion
				layout.Rigid(statusLabel.Layout)
			)
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,167 @@
+	}
+	return layout.Inset{Top: values.MarginPadding15}.Layout(gtx, func(gtx C) D {
+		return item.VoteButton.Layout(gtx)
+	})
```
```suggestion
	return layout.Inset{Top: values.MarginPadding15}.Layout(gtx, item.VoteButton.Layout)
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,167 @@
+			Top:    values.MarginPadding10,
+			Bottom: values.MarginPadding10,
+		}.Layout(gtx, text.Layout)
+	})
```
```suggestion

	return layout.Center.Layout(gtx, func(gtx C) D {
		return layout.Inset{
			Top:    values.MarginPadding10,
			Bottom: values.MarginPadding10,
		}.Layout(gtx, l.Theme.Body1("No agendas yet").Layout)
	})
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+
+func (avm *agendaVoteModal) OnDismiss() {
+
+}
```
```suggestion
func (avm *agendaVoteModal) OnDismiss() {}
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+	LiveTickets []*dcrlibwallet.Transaction
+
+	agenda               *dcrlibwallet.Agenda
+	vspIsFetched         bool
```
This is redundant.

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+			return !w.IsWatchingOnlyWallet()
+		})
+
+	avm.vspIsFetched = len(l.WL.MultiWallet.KnownVSPs()) > 0
```
Not needed we can use `len(l.WL.MultiWallet.KnownVSPs()) > 0` directly where every it is needed. Example use can be seen in the ticket purchase modal and the auto ticket modal.

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+	avm.vspSelector = components.NewVSPSelector(avm.Load).Title("Select a vsp")
+	if avm.vspIsFetched && components.StringNotEmpty(avm.WL.MultiWallet.LastUsedVSP()) {
+		avm.vspSelector.SelectVSP(avm.WL.MultiWallet.LastUsedVSP())
+	}
```
```suggestion
	lastUsedVSP := avm.WL.MultiWallet.LastUsedVSP()
	if len(avm.WL.MultiWallet.KnownVSPs()) == 0 {
		// TODO: Does this modal need this list?
		go avm.WL.MultiWallet.ReloadVSPList(context.TODO())
	} else if components.StringNotEmpty(lastUsedVSP) {
		avm.vspSelector.SelectVSP(lastUsedVSP)
	}
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+			avm.vspSelector.SelectVSP(avm.WL.MultiWallet.LastUsedVSP())
+			avm.vspIsFetched = true
+		}
+	}
```
```suggestion
	lastUsedVSP := avm.WL.MultiWallet.LastUsedVSP()
	if len(avm.WL.MultiWallet.KnownVSPs()) > 0 && lastUsedVSP != "" {
		avm.vspSelector.SelectVSP(lastUsedVSP)
	}
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+		},
+		func(gtx C) D {
+			t := avm.Theme.Body1("Select one of the options below to vote")
+			return t.Layout(gtx)
```
```suggestion
			return avm.Theme.Body1("Select one of the options below to vote").Layout(gtx)
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+	for _, k := range avm.itemKeys {
+		radioItem := layout.Rigid(avm.Load.Theme.RadioButton(avm.optionsRadioGroup, k, avm.items[k], avm.Load.Theme.Color.DeepBlue, avm.Load.Theme.Color.Primary).Layout)
+
+		items = append(items, radioItem)
```
```suggestion
radioBtn := avm.Load.Theme.RadioButton(avm.optionsRadioGroup, k, avm.items[k], avm.Load.Theme.Color.DeepBlue, avm.Load.Theme.Color.Primary)
		radioItem := layout.Rigid(radioBtn.Layout)

		items = append(items, radioItem)
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,359 @@
+				avm.Toast.NotifyError(err.Error())
+			}
+			return
+		}
```
```suggestion
                walID := avm.walletSelector.selectedWallet.ID
                vspHost := avm.vspSelector.SelectedVSP().Host
                pubKey := avm.vspSelector.SelectedVSP().PubKey
                radioBtnGrp := avm.optionsRadioGroup.Value
		err := avm.WL.MultiWallet.SetVoteChoice(walID, vspHost, pubKey, avm.agenda.AgendaID, radioBtnGrp, "", password)
		if err != nil {
			if err.Error() == dcrlibwallet.ErrInvalidPassphrase {
				avm.spendingPassword.SetError("Invalid password")
			} else {
				avm.Toast.NotifyError(err.Error())
			}
			return
		}
```

Sirmorrison commented on ui/page/governance/consensus_page.go:
```diff
@@ -0,0 +1,267 @@
+
+										return layout.Inset{Top: values.MarginPadding2}.Layout(gtx, func(gtx C) D {
+											return lastUpdatedInfo.Layout(gtx)
+										})
```
```suggestion
										return layout.Inset{Top: values.MarginPadding2}.Layout(gtx, lastUpdatedInfo.Layout)
```

Sirmorrison commented on ui/page/governance/consensus_page.go:
```diff
@@ -0,0 +1,267 @@
+								}),
+								layout.Rigid(func(gtx C) D {
+									return pg.Theme.Separator().Layout(gtx)
+								}),
```
```suggestion
								layout.Rigid(pg.Theme.Separator().Layout),
```

Sirmorrison commented on ui/page/governance/consensus_page.go:
```diff
@@ -0,0 +1,267 @@
+func (pg *ConsensusPage) layoutStartSyncSection(gtx C) D {
+	return material.Clickable(gtx, pg.syncButton, func(gtx C) D {
+		return pg.Icons.Restore.Layout24dp(gtx)
+	})
```
```suggestion
	return material.Clickable(gtx, pg.syncButton, pg.Icons.Restore.Layout24dp)
```

Sirmorrison commented on ui/page/governance/governance_page.go:
```diff
@@ -0,0 +1,314 @@
+// components unless they'll be recreated in the OnNavigatedTo() method.
+// Part of the load.Page interface.
+func (pg *Page) OnNavigatedFrom() {
+}
```
```suggestion
func (pg *Page) OnNavigatedFrom() {}
```

Sirmorrison commented on ui/page/governance/governance_page.go:
```diff
@@ -0,0 +1,314 @@
+					return layout.Flex{Axis: layout.Horizontal}.Layout(gtx,
+						layout.Rigid(func(gtx C) D {
+							return pg.layoutTabs(gtx)
+						}),
```
```suggestion
						layout.Rigid(pg.layoutTabs),
```

Sirmorrison commented on ui/page/governance/governance_page.go:
```diff
@@ -0,0 +1,314 @@
+				}),
+				layout.Rigid(func(gtx C) D {
+					return pg.Theme.Separator().Layout(gtx)
+				}),
```
```suggestion
				layout.Rigid(pg.Theme.Separator().Layout),
```

Sirmorrison commented on ui/page/governance/governance_page.go:
```diff
@@ -0,0 +1,314 @@
+								return layout.Flex{}.Layout(gtx,
+									layout.Rigid(func(gtx C) D {
+										lbl := pg.Theme.Body1(governanceTabTitles[i])
+										lbl.Color = pg.Theme.Color.Gray1
```
```suggestion
										lbl.Color = pg.Theme.Color.GrayText1
```

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -152,14 +149,8 @@ func (pg *ProposalsPage) fetchProposals() {
-
 	for pg.infoButton.Button.Clicked() {
-		pg.showInfoModal()
+		// pg.showInfoModal()
```
if no longer needed, we should remove.

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -219,70 +210,82 @@ func (pg *ProposalsPage) OnNavigatedFrom() {
-							return layout.Inset{Bottom: values.MarginPadding16}.Layout(gtx, func(gtx C) D {
-								body := func(gtx C) D {
-									return layout.Flex{Axis: layout.Vertical, Alignment: layout.End}.Layout(gtx,
+		// return components.UniformPadding(gtx, func(gtx C) D {
```
Remove if not needed.

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -219,70 +210,82 @@ func (pg *ProposalsPage) OnNavigatedFrom() {
+									// 			return components.LayoutBalanceSize(gtx, pg.Load, "678,678.687654 DCR", values.TextSize14)
+									// 		}),
+									// 	)
+									// }),
```
Remove is not needed.

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -219,70 +210,82 @@ func (pg *ProposalsPage) OnNavigatedFrom() {
+								layout.Rigid(func(gtx C) D {
+									return layout.Flex{Axis: layout.Horizontal}.Layout(gtx,
 										layout.Rigid(func(gtx C) D {
+											// return layout.Inset{Bottom: values.MarginPadding16}.Layout(gtx, func(gtx C) D {
```
Remove if not needed

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -219,70 +210,82 @@ func (pg *ProposalsPage) OnNavigatedFrom() {
+
+										return layout.Inset{Top: values.MarginPadding2}.Layout(gtx, func(gtx C) D {
+											return lastUpdatedInfo.Layout(gtx)
+										})
```
```suggestion
										return layout.Inset{Top: values.MarginPadding2}.Layout(gtx, lastUpdatedInfo.Layout)
```

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -219,70 +210,82 @@ func (pg *ProposalsPage) OnNavigatedFrom() {
+													)
 												}),
 											)
+											// })
```
Remove if not needed

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -321,11 +324,12 @@ func (pg *ProposalsPage) Layout(gtx C) D {
+				})
+			}),
+		)
+		// })
```
Remove if not needed

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -321,11 +324,12 @@ func (pg *ProposalsPage) Layout(gtx C) D {
+		)
+		// })
 	}
-	return components.UniformPadding(gtx, pg.splashScreenLayout)
```
Why was the splash screen removed?

monsa00 replied:
Is this the enable governance screen? If so, Raedah has requested we work towards removing the welcome screens to a more unified layout with on-off slider buttons for each core functionality that is not turned on by default. There are some mockups that fegor made of this. However, this was supposed to happen after 1.0 relase.

dreacot replied:
that screen is now on the governance_page.go and no longer on the proposals_page.go, as the proposals and consensus pages come under governance page

dreacot committed:
- simplified code structure
- remove unwanted comments
- cleanup code

dreacot committed:
- simplified code structure
- remove unwanted comments
- cleanup code

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,360 @@
+	"gioui.org/text"
+	"gioui.org/widget"
+	"gioui.org/widget/material"
+	"github.com/planetdecred/dcrlibwallet"
```
```suggestion

	"github.com/planetdecred/dcrlibwallet"
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,360 @@
+	LiveTickets []*dcrlibwallet.Transaction
+
+	agenda               *dcrlibwallet.Agenda
+	liveTicketsIsFetched bool
```
This seems redundant. Reason is `LiveTickets` is 0 until `avm.FetchLiveTickets()` is exec. We can simply use len(avm.LiveTickets).

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,360 @@
+
+func (avm *agendaVoteModal) FetchLiveTickets(walletID int) {
+	go func() {
+		avm.liveTicketsIsFetched = false
```
no need for this based on the comment above.

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,360 @@
+		}
+
+		avm.LiveTickets = liveTickets
+		avm.liveTicketsIsFetched = true
```
Redundant.

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,360 @@
+		avm.currentValue = avm.optionsRadioGroup.Value
+	}
+
+	if avm.liveTicketsIsFetched {
```
```suggestion
	if len(avm.LiveTickets) != 0 {
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,360 @@
+					return avm.vspSelector.Layout(gtx)
+				}),
+				layout.Rigid(func(gtx C) D {
+					if !avm.liveTicketsIsFetched {
```
```suggestion
					if len(avm.LiveTickets) == 0 {
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,162 @@
+		statusLabel = l.Theme.Label(values.MarginPadding14, agenda.Status)
+		statusLabelColor = l.Theme.Color.GreenText
+		statusIcon = decredmaterial.NewIcon(l.Icons.NavigationCheck)
+		statusIcon.Color = statusLabelColor
```
```suggestion
		statusIcon.Color = l.Theme.Color.Green500
```

lets separate icon and text colors.

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,162 @@
+		statusLabel = l.Theme.Label(values.MarginPadding14, agenda.Status)
+		statusLabelColor = l.Theme.Color.Text
+		statusIcon = decredmaterial.NewIcon(l.Icons.PlayIcon)
+		statusIcon.Color = statusLabelColor
```
```suggestion
		statusIcon.Color = l.Theme.Color.DeepBlue
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func layoutAgendaDescription(gtx C, l *load.Load, agenda dcrlibwallet.Agenda) D {
+	lbl := l.Theme.H6(agenda.Description)
```
```suggestion
	lbl :=l.Theme.Label(values.MarginPadding16, agenda.Description)
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func layoutAgendaID(gtx C, l *load.Load, agenda dcrlibwallet.Agenda) D {
+	lbl := l.Theme.H6("ID: #" + agenda.AgendaID)
```
```suggestion
	lbl := l.Theme.Label(values.MarginPadding16, "ID: #" + agenda.AgendaID)
```

Sirmorrison commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,162 @@
+			}),
+			layout.Rigid(func(gtx C) D {
+				return layoutAgendaVoteAction(gtx, l, consensusItem)
+			}),
```
`layoutAgendaStatus`, `layoutAgendaDescription`, `layoutAgendaID`, `layoutAgendaVotingPreference`, `layoutAgendaVoteAction` can all fit into one method and the value is passed to it. Reason is they all use same codes.

```
func layoutAgendaDescription(gtx C, l *load.Load, agenda dcrlibwallet.Agenda) D {
	lbl := l.Theme.H6(agenda.Description)
	lbl.Font.Weight = text.Light
	return layout.Inset{Top: values.MarginPadding4}.Layout(gtx, lbl.Layout)
}

func layoutAgendaID(gtx C, l *load.Load, agenda dcrlibwallet.Agenda) D {
	lbl := l.Theme.H6("ID: #" + agenda.AgendaID)
	lbl.Font.Weight = text.Light
	return layout.Inset{Top: values.MarginPadding4}.Layout(gtx, lbl.Layout)
}

func layoutAgendaVotingPreference(gtx C, l *load.Load, agenda dcrlibwallet.Agenda) D {
	lbl := l.Theme.H6("Voting Preference: " + agenda.VotingPreference)
	lbl.Font.Weight = text.Light
	return layout.Inset{Top: values.MarginPadding4}.Layout(gtx, lbl.Layout)
}
```

dreacot replied:
@Sirmorrison the current implementation enhances code readability

Sirmorrison reviewed (changes requested):
<img width="799" alt="image" src="https://user-images.githubusercontent.com/27733432/153749291-d7e89a50-d5ed-4dca-90ac-7ca999436954.png">
* The vote/change vote button text is wrapped and not properly aligned.
* Every word on the agenda subject/title should be capitalized.

<img width="405" alt="image" src="https://user-images.githubusercontent.com/27733432/153749403-d81ec660-02fe-4f6c-accb-0e2e6d45ef98.png">

The vote button should not be visible for finished vote.
Also, use a shadow widget for the agendas when they are hovered so the background is not hidden when hovered.
Something similar was done for the more page and the security page.
<img width="576" alt="image" src="https://user-images.githubusercontent.com/27733432/153750372-d38a910a-c960-496a-bdf6-5b564be4a1f4.png">

The navigation does not conform to the mock up.
<img width="609" alt="image" src="https://user-images.githubusercontent.com/27733432/153749525-da4bc975-e534-438c-a863-4a5ff9672a5d.png">

mock up 
<img width="424" alt="image" src="https://user-images.githubusercontent.com/27733432/153749571-4488c8d0-0073-46aa-a513-2a345fc9f997.png">

dreacot committed:
- WIP: add hover effect to icon on consensus page

dreacot committed:
- WIP: add hover effect to icon on consensus page

dreacot committed:
fix vote button wrapping

dreacot committed:
remove redundant boolean call

dreacot committed:
remove redundant boolean call

dreacot committed:
change H6 to Label

dreacot committed:
add shadow to agenda items

dreacot committed:
add open in browser section

dreacot committed:
add open in browser section

Sirmorrison committed:
Rearrange widget position, label font size and code clean up

Sirmorrison committed:
implement shadow on clickable_list widget
- remove redundant lines of codes
- fix excess padding and general code clean up

Sirmorrison committed:
move consenses and proposal handle properties to individual pages.
- clean up and optimize code

dreacot committed:
- some modifications were made to dcrlibwallet on how some consensus methods were called, the necessary changes were implemented in this commit; mainly changing the consensus methods from multiwallet methods to wallet methods

dreacot committed:
- show modal when the info icons on the governance page is clicked
 - make the voting dashboard redirect to the correct url
 - fix crash that occured when navigating from the governance page

dreacot committed:
update dcrlibwallet dep

dreacot committed:
update dcrlibwallet dep

itswisdomagain reviewed (changes requested):
First pass. General feedback from testing.

- Do you really need to have live tickets to be able to set voting preferences? Remember that locally saved preferences are applied to new tickets, so I suppose it is fine to let users save preferences locally only, if there's no ticket purchased yet.
- Governance page -> Proposals tab shows one prop in discussion. Decrediton has none.
- Governance page -> Consensus tab shows 4 in progress agendas, Decrediton has none in progress. Also, the link to the voting dashboard for testnet should be `https://voting.decred.org/testnet`.

Lastly, and this may be something for after this PR is merged. The green tick icon for finished agendas is meant to show that the agenda voting passed. Decrediton uses dcrdata API to get this info.

itswisdomagain commented on ui/page/components/components.go:
```diff
@@ -714,3 +733,122 @@ func CoinImageBySymbol(icons *load.Icons, coinName string) *decredmaterial.Image
 	return nil
 }
+
+func WalletLiveTickets(w *dcrlibwallet.Wallet) ([]dcrlibwallet.Transaction, error) {
```
This seems like a candidate for a dcrlibwallet `wallet.UnspentTickets` method. Unspent, not Live because unmined tickets are not live but I see they are returned by this function as well.

dreacot replied:
can you please link the method, can't find it

itswisdomagain replied:
Which method? I mean to move this `WalletLiveTickets` method to dcrlibwallet and call it 
~`wallet.UnspentTickets`~ `wallet.UnspentUnexpiredTickets` which better represents the kind of tickets returned by the method.

itswisdomagain commented on ui/page/components/components.go:
```diff
@@ -714,3 +733,122 @@ func CoinImageBySymbol(icons *load.Icons, coinName string) *decredmaterial.Image
+	return tickets, nil
+}
+
+func StakeToTransactionItems(l *load.Load, txs []dcrlibwallet.Transaction, newestFirst bool, hasFilter func(int32) bool) ([]*TransactionItem, error) {
```
Please document this function, don't want to have to read through to understand what it does. It should be a general practice to document all public functions, methods and properties.

dreacot replied:
i didn't author that function, it was originally in the staking utils file(https://github.com/planetdecred/godcr/pull/729/files#diff-645fdc2d09679835c6acde6c7a80862d9375e5eede766ab6775127e8863cd28cL59), i only moved it to components

itswisdomagain replied:
I suppose you called this function, so you have a good idea of what its purpose is. Please document to the best of your understanding. @Sirmorrison can help review the comment afterwards to confirm it matches the function's purpose.

itswisdomagain commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,145 @@
+	VoteButton decredmaterial.Button
+}
+
+func AgendasList(gtx C, l *load.Load, consensusItem *ConsensusItem) D {
```
Does this return a list? Looks more like a `AgendaItemWidget` to me.

itswisdomagain commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,145 @@
+	var backgroundColor color.NRGBA
+
+	switch agenda.Status() {
+	case "Finished":
```
Define these as constants in dcrlibwallet and use here instead of string literals. `AgendaStatusFinished = "Finished"` for example.

itswisdomagain commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,145 @@
+
+func layoutAgendaVoteAction(gtx C, l *load.Load, item *ConsensusItem) D {
+	gtx.Constraints.Min.X, gtx.Constraints.Max.X = gtx.Px(unit.Dp(150)), gtx.Px(unit.Dp(150))
+	if canVote {
```
You don't need a public canVote property that is set in one function and used in a different function.
```suggestion
	if item.Agenda.Status() == "In progress" { // remember to use a dcrlibwallet constant instead
```

itswisdomagain reviewed (changes requested):
Haven't looked too deep, but here's for now.

dreacot committed:
add testnest host for the voting dashboard

dreacot committed:
remove vsp selector from the agenda vote modal
update  with dcrlibwallet

dreacot committed:
- rename AgendaList to AgendaItemWidget
- use dcrlibwallet agenda status constants

dreacot committed:
update dcrlibwallet dep

dreacot committed:
- move WalletLiveTickets() method to dcrlibwallet
- document StakeToTransactionItems()
- update dcrlibwallet dep
- change vote text on agenda button"

dreacot committed:
fix update preference button width

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,302 @@
+			avm.loadCount++
+			avm.ticketSelector = newTicketSelector(avm.Load, avm.LiveTickets).Title("Select a ticket")
+		}
+
```
Remove white space.

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,302 @@
+		},
+		func(gtx C) D {
+			return avm.walletSelector.Layout(gtx)
+		},
```
```suggestion
		avm.Theme.Body1("Select one of the options below to vote").Layout,
		avm.walletSelector.Layout,
```

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,302 @@
+					return avm.spendingPassword.Layout(gtx)
+				}),
+			)
+		},
```
```suggestion
		func(gtx C) D {
			return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
				layout.Rigid(avm.spendingPassword.Layout),
			)
		},
```

Sirmorrison commented on ui/page/governance/proposals_page.go:
```diff
@@ -218,114 +214,51 @@ func (pg *ProposalsPage) OnNavigatedFrom() {
-									})
+								return layout.UniformInset(values.MarginPadding8).Layout(gtx, func(gtx C) D {
+									return pg.layoutSyncSection(gtx)
 								})
```
```suggestion
								return layout.UniformInset(values.MarginPadding8).Layout(gtx, func(gtx C) D {
									return pg.layoutSyncSection(gtx)
								})
```
```suggestion
								return layout.UniformInset(values.MarginPadding8).Layout(gtx, pg.layoutSyncSection)
```

Sirmorrison reviewed (changes requested):
I have also noticed that on the agenda vote modals, the radio buttons options tend to shuffle as against being statically placed.
![ezgif com-gif-maker(4)](https://user-images.githubusercontent.com/27733432/157193346-cff4afab-98ca-468f-9373-972d0f079e4c.gif)

dreacot committed:
fix bug where voting options were being re-arranged
cleanup code

dreacot committed:
update agenda status call

itswisdomagain commented on ui/decredmaterial/clickable_list.go:
```diff
@@ -53,6 +59,11 @@ func (cl *ClickableList) handleClickables(count int) {
+			return cl.shadow.Layout(gtx, func(gtx C) D {
+				return cl.row(gtx, count, i, w)
+			})
+		}
```
You don't need to store a `cl.shadow` property.
```suggestion
		if cl.IsShadowEnabled && cl.clickables[i].button.Hovered() {
			shadow = cl.theme.Shadow()
			shadow.SetShadowRadius(14)
			shadow.SetShadowElevation(5)
			return shadow.Layout(gtx, func(gtx C) D {
				return cl.row(gtx, count, i, w)
			})
		}
```

itswisdomagain commented on ui/page/components/components.go:
```diff
@@ -23,6 +24,24 @@ import (
 	"github.com/planetdecred/godcr/ui/values"
 )
 
+type TransactionItem struct {
```
Needs a comment.

itswisdomagain commented on ui/page/components/components.go:
```diff
@@ -714,3 +733,108 @@ func CoinImageBySymbol(icons *load.Icons, coinName string) *decredmaterial.Image
 }
+
+// StakeToTransactionItems converts stake transactions to an array of TransactionItem
+func StakeToTransactionItems(l *load.Load, txs []dcrlibwallet.Transaction, newestFirst bool, hasFilter func(int32) bool) ([]*TransactionItem, error) {
```
I want to recommend undoing this change and keeping as it was without touching any of it. It's quite complex and not properly commented. I've identified one new area in this PR that relies on this function and I'm proposing an alternative below. If there are other new areas that require/use this function, let's discuss how to do without this function in those areas.

dreacot replied:
you mean undoing as in removing the comment, or undoing as taking the method back to the file it was in before?

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+			liveTickets = append(liveTickets, liveTicket.Transaction)
+		}
+
+		avm.LiveTickets = liveTickets
```
You don't need these. It seems the `StakeToTransactionItems` was designed to deal with voted and expired tickets which we don't have here. Only other sensible thing it does is sort.
```suggestion
		// sort by newest first
		sort.Slice(tickets[:], func(i, j int) bool {
			var timeStampI = tickets[i].Timestamp
			var timeStampJ = tickets[j].Timestamp
			return timeStampI > timeStampJ
		})
		avm.LiveTickets = make([]*dcrlibwallet.Transaction, len(tickets))
		for i := range tickets {
			avm.LiveTickets[i] = &tickets[i]
		}
```

itswisdomagain commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,140 @@
+
+func AgendaItemWidget(gtx C, l *load.Load, consensusItem *ConsensusItem) D {
+	gtx.Constraints.Min.X = gtx.Constraints.Max.X
+	agenda := consensusItem.Agenda
```
Just use `consensusItem.Agenda` directly.
```suggestion
```

itswisdomagain commented on ui/page/components/consensus_list.go:
```diff
@@ -0,0 +1,140 @@
+			consensusItems = append(consensusItems, item)
+		}
+	}
+	return consensusItems
```
```suggestion
	agendas, err := selectedWallet.AllVoteAgendas("", newestFirst)
	if err != nil {
		return nil
	}

	consensusItems := make([]*ConsensusItem, len(agendas))
	for i := 0; i < len(agendas); i++ {
		consensusItems[i] = &ConsensusItem{
			Agenda:     *agendas[i],
			VoteButton: l.Theme.Button("Update Preference"),
		}
	}
	return consensusItems
```

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+	*load.Load
+	modal decredmaterial.Modal
+
+	LiveTickets []*dcrlibwallet.Transaction
```
Not used outside this package. Also, these tickets aren't exactly live, let's call them votable, and add a comment explaining what that means.
```suggestion
	votableTickets []*dcrlibwallet.Transaction // tickets that have not been spent by a vote or revocation (unspent) and that have not expired (unexpired).
```

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+	isVoting  bool
+	loadCount int
+
+	consensusPage *ConsensusPage
```
Do not link to a page, use a callback instead to execute a method on the calling page whenever necessary here. Compare a callback is used in the create password modal. https://github.com/planetdecred/godcr/blob/c66d74aaee17c798bc64b976c61bc43edb7eae67/ui/modal/create_password_modal.go#L136

dreacot replied:
https://github.com/planetdecred/godcr/pull/729/files#diff-c857bf384d3c25f21adf5db9d32b10a475d3fcee66adaa7d3620eb61518227bbR253, the page liking was used to refresh the agendas list. Currently not seeing the feasibility on linking the consensus page and agenda vote modal using a callback

can you push a commit to this effect

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+	itemKeys          []string
+	defaultValue      string // str-key
+	initialValue      string
+	currentValue      string
```
What items, what values, what keys? These are extremely generic, can never deduce what purpose these variables serve from just looking at the variable names and types.

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+	avm.voteBtn.Color = l.Theme.Color.Surface
+
+	avm.walletSelector = NewWalletSelector(l).
+		Title("Voting wallet").
```
Not sure what you mean by "Voting wallet". Godcr wallets cannot vote. Only vspd wallets. Maybe just "Select wallet"?

dreacot replied:
this is meant to refer to the wallets selected to update an agenda preference

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+	avm.walletSelector = NewWalletSelector(l).
+		Title("Voting wallet").
+		WalletSelected(func(w *dcrlibwallet.Wallet) {
+			avm.loadCount = 0
```
A count of loads?? This variable name can use some love.

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+			avm.loadCount = 0
+
+			avm.FetchUnspentUnexpiredTickets(w.ID)
+			avm.RefreshWindow()
```
This shouldn't be necessary. The modal is dismissed after this function ends, and that triggers a window refresh.

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+							return stringI < stringJ
+						}
+						return stringI < stringJ
+					})
```
If you use an array instead of a map, you shouldn't need to sort. Also, the key is the same as the value, even though read from different objects; because `agenda` is the same as `consensusItem.Agenda`.
`ArrVoteOptions := make(map[string]string)` should be `voteChoices := make([]string, len(consensusItem.Agenda.Choices))`

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,307 @@
+					initialValue := consensusItem.Agenda.VotingPreference
+					if initialValue == "" {
+						initialValue = avm.defaultValue
+					}
```
`avm.defaultValue` is never assigned a value and so is basically the same as "". The field can be removed. Also, `initialValue` => `currentVotePref`.

itswisdomagain reviewed (changes requested):
Are the icons added here provided by the design team? Or more like placeholders pending when the design team provides something?

Will re-review after these.

dreacot committed:
- don't store cl.shadow property
- cleanup code

dreacot committed:
- give variables in agenda modal proper names
 - change "Voting wallet" to selected "wallet"
 - use an array instead of a map for the vote choices

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,255 @@
+	isVoting         bool
+	modalUpdateCount int // this keeps track of the number of times the modal has been updated.
+
+	consensusPage *ConsensusPage
```
```suggestion
	onDismiss func()
```

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,255 @@
+	cancelBtn         decredmaterial.Button
+}
+
+func newAgendaVoteModal(l *load.Load, agenda *dcrlibwallet.Agenda, consensusPage *ConsensusPage) *agendaVoteModal {
```
```suggestion
func newAgendaVoteModal(l *load.Load, agenda *dcrlibwallet.Agenda, onDismiss func()) *agendaVoteModal {
```

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,255 @@
+		Load:              l,
+		modal:             *l.Theme.ModalFloatTitle(),
+		agenda:            agenda,
+		consensusPage:     consensusPage,
```
```suggestion
		onDismiss:     onDismiss,
```

itswisdomagain commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,255 @@
+		avm.Toast.Notify("Vote preference updated successfully")
+
+		avm.Dismiss()
+		go avm.consensusPage.FetchAgendas()
```
```suggestion
		avm.onDismiss()
```

itswisdomagain commented on ui/page/governance/consensus_page.go:
```diff
@@ -0,0 +1,307 @@
+
+	for i := range pg.consensusItems {
+		if pg.consensusItems[i].VoteButton.Clicked() {
+			newAgendaVoteModal(pg.Load, &pg.consensusItems[i].Agenda, pg).Show()
```
```suggestion
			newAgendaVoteModal(pg.Load, &pg.consensusItems[i].Agenda, func() {
				go pg.FetchAgendas() // re-fetch agendas when modal is dismissed
			}).Show()
```

itswisdomagain reviewed (changes requested):
I'm not too sure about the name `onDismiss` since the function isn't always called when the modal is dismissed but here's the idea for using a callback.

dreacot committed:
- use call back to reload agenda list after updating the voting preference

Sirmorrison commented on ui/page/governance/agenda_vote_modal.go:
```diff
@@ -0,0 +1,254 @@
+	*load.Load
+	modal decredmaterial.Modal
+
+	votableTickets []*dcrlibwallet.Transaction // tickets that have not been spent by a vote or revocation (unspent) and that have not expired (unexpired).
```
```suggestion
        // tickets that have not been spent by a vote or revocation (unspent) and that
        // have not expired (unexpired).
	votableTickets []*dcrlibwallet.Transaction 
```

dreacot committed:
move stakeToTransactionItems back to staking utils page

itswisdomagain reviewed (changes requested):
Would have left review comments but did some tinkering while testing and want to share the thoughts in this commit instead: https://github.com/itswisdomagain/godcr-1/commit/4d1dfeb700dca1b9daee3f9399dc163c6eb3c64c.

That should resolve the slowness issue @Sirmorrison experienced, must have been due to the network call being made on the UI thread.

Also bump the dcrlibwallet mod version to master once the related dcrlibwallet PR is merged.

dreacot committed:
fix delay while opening consensus tab
code cleanup

dreacot committed:
bump dcrlibwallet dep

dreacot committed:
allow users vote for upcoming agendas

itswisdomagain commented on go.mod:
```diff
@@ -17,9 +18,9 @@ require (
 	golang.org/x/image v0.0.0-20210628002857-a66eb6448b8d
 	golang.org/x/text v0.3.7
-)
+)
```
Need to run go mod tidy. That will restore the newline at the end of this file and also fix the indirect imports in the file.

itswisdomagain reviewed (dismissed):
I changed the status of one of the agendas to upcoming in order to test how updating the voting preference works, noticed this slight issue (may not even be an issue).

When my wallet launched, it showed I had 2 immature tickets, although they had actually voted but my wallet isn't synced yet, so it doesn't know. When I tried to update voting preferences for the agenda I modified to upcoming, it gave an error about `ticket already voted`. The issue resolved itself when the wallet finished syncing.

I'm currently unable to buy another ticket to test with, to ensure the vsp associated with the ticket gets the correct voting preference set. But this should be good to merge after the go mod tidy stuff below. There'd be more testing to be done after this hits master, such as ensuring that tickets purchased before and after setting the vote preference gets the correct preference set with the vsp.

dreacot committed:
cleanup code
